File: apps/backend/src/user/user.service.ts
import { Injectable, NotFoundException } from "@nestjs/common"
import { InjectRepository } from "@nestjs/typeorm"
import { Repository } from "typeorm"

import { CreateUserDto, UpdateUserDto } from "./user.dto"
import { User } from "./user.entity"

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>
  ) {}

  async getUserByEmail(email: string): Promise<User> {
    const result = await this.userRepository.findOne({
      where: { email },
    })
    return result
  }

  async getUserById(id: number): Promise<User> {
    const result = await this.userRepository.findOne({
      where: { id },
    })
    return result
  }

  async findByEmailOrSave(email, user_name, provider_id): Promise<User> {
    const foundUser = await this.getUserByEmail(email)
    if (foundUser) {
      return foundUser
    }

    return await this.userRepository.save({
      user_name,
      email,
      provider_id,
      provider: "google",
    })
  }

  async updateNickname(
    id: number,
    userData: UpdateUserDto
  ): Promise<User | null> {
    const result = await this.userRepository.update(id, {
      user_name: userData.username,
    })

    if (result.affected > 0) {
      return this.userRepository.findOne({ where: { id } })
    }

    throw new NotFoundException(`ID ${id}가 잘못되었습니다.`)
  }
}

File: apps/backend/kotlinboot/src/main/kotlin/com/soulfriends/kotlinboot/user/UserService.kt
package com.soulfriends.kotlinboot.user

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(private val userRepository: UserRepository) {

    @Transactional
    fun createUser(email: String, providerId: String, provider: String): User {
        return userRepository.save(User(email = email, providerId = providerId, provider = provider))
    }

    fun getUserByEmail(email: String): User? {
        return userRepository.findByEmail(email)
    }

    fun getUserById(id: Long): User? {
        return userRepository.findById(id).orElse(null)
    }

    @Transactional
    fun findByEmailOrSave(email: String, providerId: String, provider: String): User {
        return getUserByEmail(email) ?: createUser(email, providerId, provider)
    }

    fun checkNickname(userName: String): Boolean {
        return userRepository.findByUserName(userName) == null
    }

    @Transactional
    fun updateNickname(id: Long, userName: String): User? {
        val user = getUserById(id) ?: return null
        user.userName = userName
        return userRepository.save(user)
    }

    fun findUserByRefreshToken(refreshToken: String): User? {
        return userRepository.findByRefreshToken(refreshToken)
    }
}
